src/transformers/modeling_dpr.py

```diff
@@ -0,0 +1,541 @@
+
+
+class DPRSpanPredictor(PreTrainedModel):
+    def __init__(self, config: DPRConfig):
+        super().__init__(config)
+        self.encoder = DPREncoder(config)
+        self.qa_outputs = nn.Linear(self.encoder.embeddings_size, 2)
+        self.qa_classifier = nn.Linear(self.encoder.embeddings_size, 1)
+        self.init_weights()
+
+    def forward(self, input_ids: Tensor, attention_mask: Tensor):
+        # notations: N - number of questions in a batch, M - number of passages per questions, L - sequence length
+        N, M, L = input_ids.size()
+        # resize as matrices
+        input_ids = input_ids.view(N * M, L)
+        attention_mask = attention_mask.view(N * M, L)
+        # feed encoder
+        sequence_output, *_ = self.encoder(input_ids, None, attention_mask)
+        # compute logits
+        logits = self.qa_outputs(sequence_output)
+        start_logits, end_logits = logits.split(1, dim=-1)
+        start_logits = start_logits.squeeze(-1)
+        end_logits = end_logits.squeeze(-1)
+        relevance_logits = self.qa_classifier(sequence_output[:, 0, :])
+        # resize and return
+        return start_logits.view(N, M, L), end_logits.view(N, M, L), relevance_logits.view(N, M)
+
+    def init_weights(self):
+        self.encoder.init_weights()
+
+
+##################
+# PreTrainedModel
+##################
+
+
+CheckpointState = collections.namedtuple(
+    "CheckpointState", ["model_dict", "optimizer_dict", "scheduler_dict", "offset", "epoch", "encoder_params"]
+)
+
+
+def load_states_from_checkpoint(model_file: str) -> CheckpointState:
+    logger.info("Reading saved model from %s", model_file)
+    state_dict = torch.load(model_file, map_location=lambda s, l: default_restore_location(s, "cpu"))
+    logger.info("model_state_dict keys %s", state_dict.keys())
+    return CheckpointState(**state_dict)
+
+
+class DPRPretrainedContextEncoder(PreTrainedModel):
+    """ An abstract class to handle weights initialization and
+        a simple interface for downloading and loading pretrained models.
+    """
+
+    config_class = DPRConfig
+    load_tf_weights = None
+    base_model_prefix = "ctx_encoder"
+
+    def init_weights(self):
+        """Load the weights from the official DPR repo's format if specified."""
+        if self.config.bi_encoder_model_file is not None:
+            logger.info("Loading DPR biencoder from {}".format(self.config.bi_encoder_model_file))
+            saved_state = load_states_from_checkpoint(self.config.bi_encoder_model_file)
+            encoder, prefix = self.ctx_encoder, "ctx_model."
+            state_dict = {}
+            for key, value in saved_state.model_dict.items():
+                if key.startswith(prefix):
+                    key = key[len(prefix) :]
+                    if not key.startswith("encode_proj."):
+                        key = "bert_model." + key
+                    state_dict[key] = value
+            encoder.load_state_dict(state_dict)
```

Comment:
This is very different from how we handle weights loading in the rest of the library (using `init_weights` for random initialization and `from_pretrained` for loading pretrained weights).

So usually we have a conversion script (all the `convert_xxx.py`) which allows us to convert the author provided checkpoints in the library checkpoint (`state_dict`). We usually take this occasion to remove the optimizer averages and all the annex weights to reduce the checkpoint sizes.

Is there a specific reason to have such a different logic here? Otherwise, I think we should take the same approach as in the rest of the library.

Reply:
I changed it to a convert script